defrag/defrag.go

```diff
@@ -0,0 +1,276 @@
+package defrag
+
+import (
+	"container/list"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"code.google.com/p/gopacket"
+	"code.google.com/p/gopacket/layers"
+)
+
+// Quick and Easy yo use debug code to trace
+// how defrag works.
+var debug debugging = false // or flip to false
+type debugging bool
+
+func (d debugging) Printf(format string, args ...interface{}) {
+	if d {
+		log.Printf(format, args...)
+	}
+}
+
+type Defragmenter interface {
+	// DefragIPv4 takes in an IPv4 packet with a fragment payload.
+	//
+	// It modifies the IPv4 layer in place, returning true if the
+	// modified layer is now a full IPv4 payload.
+	//
+	// If the passed-in IP layer is NOT fragmented, it will
+	// immediately return true without modifying the layer.
+	//
+	// If the IPv4 layer is a fragment and we don't have all
+	// fragments, it will return false and store whatever internal
+	// information it needs to eventually defrag the packet.
+	//
+	// If the IPv4 layer is the last fragment needed to reconstruct
+	// the packet, *ip will be set to the entire defragmented packet,
+	// and the function will return true.
+	DefragIPv4(IPv4 *layers.IPv4) (bool, error)
```

Comment:
Here's the code I'm envisioning for the defragger:

    func (d *IPv4) DefragIPv4(in *layers.IPv4) (*layers.IPv4, error) {
      if err := checkValidity(in); err != nil {
        return nil, err
      } else if notFragmented(in) {
       return in, nil
      }
      packetDefragger := d.findDefragger(in.src, in.dst, in.id)
      if err := packetDefragger.addFragment(in); err != nil {
        return nil, err
      }
      if packetDefragger.complete() {
        return packetDefragger.defragged(), nil
      }
      return nil, nil
    }

And here's the code I'm envisioning for the client:

    func HandlePacket(in *layers.IPv4) err {
      in, err := defragger.DefragIPv4(in)
      if err != nil {
        return err
      } else if in == nil {
        return nil  // packet fragment, we don't have whole packet yet.
      }
      // At this point, we know that 'in' is defragmented.  It may be the same 'in' passed to
      // HandlePacket, or it may not, but we don't really care :)
      ... do stuff to 'in' ...
    }